katharsis-core/src/main/java/io/katharsis/resource/registry/repository/adapter/ResponseRepositoryAdapter.java

```diff
@@ -0,0 +1,348 @@
+ * a resource object or a {@link JsonApiResponse} response which should be returned by a controller. Ok, the last
+ * sentence is not 100% true since interface based repositories can return only resources, but who's using it anyway?
+ * <p>
+ * The methods need to know if a repository is interface- or annotation-based since repository methods have different
+ * signatures.
+ */
+public abstract class ResponseRepositoryAdapter {
+
+	protected ResourceInformation resourceInformation;
+
+    protected ModuleRegistry moduleRegistry;
+
+    public ResponseRepositoryAdapter(ResourceInformation resourceInformation, ModuleRegistry moduleRegistry) {
+        this.moduleRegistry = moduleRegistry;
+        this.resourceInformation = resourceInformation;
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+	protected <T> Iterable<T> filterResult(Iterable<?> resources, RepositoryRequestSpec requestSpec) {
+    	RepositoryResultFilterChainImpl<T> chain = new RepositoryResultFilterChainImpl<>((Iterable)resources);
+    	return chain.doFilter(newRepositoryFilterContext(requestSpec));
+    }
+    
+	protected JsonApiResponse getResponse(Object repository, Object result, RepositoryRequestSpec requestSpec) {
+        if (result instanceof JsonApiResponse) {
+            return (JsonApiResponse) result;
+        }
+
+        Iterable<?> resources;
+        boolean isCollection = result instanceof Iterable;
+        if (isCollection) {
+            resources = (Iterable<?>) result;
+        } else {
+            resources = Collections.singletonList(result);
+        }
+        Iterable<?> filteredResult = filterResult(resources, requestSpec);
+        MetaInformation metaInformation = getMetaInformation(repository, resources, requestSpec);
+        LinksInformation linksInformation = getLinksInformation(repository, resources, requestSpec);
+
+        Object resultEntity;
+        if(isCollection){
+        	resultEntity = filteredResult;
+        }else {
+        	Iterator<?> iterator = filteredResult.iterator();
+        	if(iterator.hasNext()){
+        		resultEntity = iterator.next();
+        		PreconditionUtil.assertFalse("expected unique result", iterator.hasNext());
+        	}else{
+        		resultEntity = null;
+        	}
+        }
+        
+        return new JsonApiResponse().setEntity(resultEntity).setLinksInformation(linksInformation)
+                .setMetaInformation(metaInformation);
+    }
+
+    private MetaInformation getMetaInformation(Object repository, Iterable<?> resources, RepositoryRequestSpec requestSpec) {
+    	RepositoryMetaFilterChainImpl chain = new RepositoryMetaFilterChainImpl(repository);
+    	return chain.doFilter(newRepositoryFilterContext(requestSpec), resources);
+    }
+    
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private MetaInformation doGetMetaInformation(Object repository, Iterable<?> resources, RepositoryRequestSpec requestSpec) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Move this method into "RepositoryMetaFilterChainImpl". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS3398)